### ISSUE TITLE
The "ResolvePackageAssets" task failed unexpectedly due to max path limitation of NuGet package cache

### ISSUE THREAD

**Issue by JamesNK** [dotnet/sdk#2044]
I am using version `2.1.300-preview2-008251` of the SDK and I am getting this failure multiple times when building inside VS. Building from the command line completes successfully.

```
Severity	Code	Description	Project	File	Line	Suppression State
Error	MSB4018	The "ResolvePackageAssets" task failed unexpectedly.
System.ArgumentNullException: Value cannot be null.
Parameter name: path1
   at System.IO.Path.Combine(String path1, String path2)
   at Microsoft.NET.Build.Tasks.ResolvePackageAssets.RaisePackageAssets[T](LockFileTarget target, Func`2 getAssets, Func`2 filter, Action`2 setup)
   at Microsoft.NET.Build.Tasks.ResolvePackageAssets.ExecuteCore()
   at Microsoft.NET.Build.Tasks.TaskBase.Execute()
   at Microsoft.Build.BackEnd.TaskExecutionHost.Microsoft.Build.BackEnd.ITaskExecutionHost.Execute()
   at Microsoft.Build.BackEnd.TaskBuilder.<ExecuteInstantiatedTask>d__26.MoveNext()	SocialWeather	C:\Users\James\.dotnet\x64\sdk\2.1.300-preview2-008251\Sdks\Microsoft.NET.Sdk\build\Microsoft.PackageDependencyResolution.targets	201	
```

Failing source: https://github.com/aspnet/SignalR/commit/4b666dfa856a7600205ab488a7ad4c8facbe7f6e

Open SignalR.sln in 15.6.2 and rebuild solution for the error.

[buildlogs.zip](https://github.com/dotnet/sdk/files/1812858/buildlogs.zip)

**Comment by livarcocc** [comment]
@nguerrera you said you were going to take a look at it. So, assigning to you for now.

**Comment by JamesNK** [comment]
Hi

What is the ETA for this to be looked at? It is super urgent but it is blocking me from working on an issue

**Comment by nguerrera** [comment]
I will look at it tomorrow morning. I had intended to get to it right away and got interrupted.

**Comment by nguerrera** [comment]
Looking right now. Sorry for delay.

**Comment by nguerrera** [comment]
first issue I hit is different:

Restore fails with: `Error occurred while restoring NuGet packages: The specified path, file name, or both are too long. The fully qualified file name must be less than 260 characters, and the directory name must be less than 248 characters.`

Build then can't find assets files.

cc @rrelyea @nkolev92

**Comment by nguerrera** [comment]
Building on command line first gets passed that, but then build succeeds:

```
========== Build: 30 succeeded, 0 failed, 0 up-to-date, 0 skipped ==========
```

I will check the binlog

**Comment by nguerrera** [comment]
OK, log reveals critical part of repro: `Rebuild`.

Workaround: Use git clean -fdx to clean and only use `Build`. Still looking for the bug.

**Comment by nguerrera** [comment]
The catalyst seems to be that
```
C:\Users\nicholg\.nuget\packages\microsoft.aspnetcore.authentication.abstractions\2.1.0-a-preview2-clean-up-proto-abstractions-17457\microsoft.aspnetcore.authentication.abstractions.2.1.0-a-preview2-clean-up-proto-abstractions-17457.nupkg.sha512
```

is perilously close to MAX_PATH.

Restore target fails when running in desktop msbuild:

```
D:\S\SignalR.sln" (restore target) (1) ->
(Restore target) ->
  C:\Program Files (x86)\Microsoft Visual Studio\Preview\Enterprise\Common7\IDE\CommonExtensions\Microsoft\NuGet\NuGet.targets(104,5): error
: Could not find a part of the path 'microsoft.aspnetcore.server.kestrel.transport.abstractions.2.1.0-a-preview2-clean-up-proto-abstractions- 17574.nupkg.sha512'
```

I can harden the task to report a better error message, but that won't unblock you. I think this is either an issue in msbuild or nuget. cc @AndyGerlicher  @rainersigwald and @rrelyea @nkolev92 again. It's going to take a while to sort out?

Can you shorten that comical version number?

**Comment by anurse** [comment]
cc @natemcmaster @pranavkm 

This is due to our "feature branch" versions :(

**Comment by natemcmaster** [comment]
Would reducing the nesting of the obj/ and bin/ folders help? We could try overriding BaseIntermediateOutputPath and BaseOutputPath

**Comment by natemcmaster** [comment]
oh wait, nvm, it seems ~/.nuget/packages is the root...

**Comment by nguerrera** [comment]
> it is blocking me from working on an issue

@JamesNK See the (edited) workaround here: https://github.com/dotnet/sdk/issues/2044#issuecomment-373544638

Is that enough to consider that you are at least unblocked from working on that issue?

**Comment by nguerrera** [comment]
I will fix the error message to say something like "Package 'X, version Y' was not found. It may have been deleted since restore.  Otherwise, restore may have been only partially completed, which may have been due to path length restrictions."

**Comment by JamesNK** [comment]
> Is that enough to consider that you are at least unblocked from working on that issue?

Code has been updated to not use the problem package. I'm not blocked anymore.

**Comment by nguerrera** [comment]
Fixed by #2210
